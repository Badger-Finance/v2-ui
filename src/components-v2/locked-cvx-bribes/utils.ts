import BigNumber from 'bignumber.js';
import { formatWithoutExtraZeros, numberWithCommas, unscale } from '../../mobx/utils/helpers';

export function formatBalance(balance: BigNumber): string {
  return numberWithCommas(formatWithoutExtraZeros(unscale(balance, 18)));
}
