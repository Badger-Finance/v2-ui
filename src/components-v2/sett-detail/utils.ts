import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/setts/sett-charts';
import { DelaySeverity } from '../../mobx/model/setts/sett-rewards';
import { SettDetailMode } from '../../mobx/model/setts/sett-detail';

dayjs.extend(utc);

export const ChartModeTitles = {
	[ChartMode.value]: 'Sett Value',
	[ChartMode.ratio]: 'Token Ratio',
	[ChartMode.accountBalance]: 'Your Holdings',
};

export const SettModeTitles = {
	[SettDetailMode.settInformation]: 'Vault Info',
	[SettDetailMode.userInformation]: 'User Info',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};
