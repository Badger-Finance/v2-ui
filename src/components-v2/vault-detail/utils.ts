import { Vault, VaultData } from '@badger-dao/sdk';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/vaults/vault-charts';
import { DelaySeverity } from '../../mobx/model/vaults/vault-rewards';

dayjs.extend(utc);

export const ChartModeTitles: Record<string, string> = {
	[ChartMode.Value]: 'Sett Value',
	[ChartMode.Ratio]: 'Token Ratio',
	[ChartMode.AccountBalance]: 'My Holdings',
	[ChartMode.BoostMultiplier]: 'Badger Boost',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};

export function defaultVaultBalance(vault: Vault): VaultData {
	return {
		address: vault.vaultToken,
		name: vault.name,
		symbol: vault.asset,
		pricePerFullShare: vault.pricePerFullShare,
		balance: 0,
		value: 0,
		earnedBalance: 0,
		earnedValue: 0,
		depositedBalance: 0,
		withdrawnBalance: 0,
		tokens: [],
		earnedTokens: [],
	};
}

export interface Balance {
	balance: number;
}

export function hasBalance(input?: Balance): boolean {
	if (!input) {
		return false;
	}
	return input.balance > 0;
}
