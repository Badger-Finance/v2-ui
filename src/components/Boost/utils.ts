import { makeStyles } from '@material-ui/core/styles';
import { ClassNameMap } from '@material-ui/core/styles/withStyles';
import React from 'react';

type ComparisonConfig = {
  toCompareValue: number;
  toBeComparedValue: number;
  greaterCaseColor: string;
  lessCaseColor: string;
  defaultColor: string;
};

export const getColorFromComparison = ({
  toCompareValue,
  toBeComparedValue,
  greaterCaseColor,
  lessCaseColor,
  defaultColor,
}: ComparisonConfig): string => {
  if (toCompareValue > toBeComparedValue) return greaterCaseColor;
  if (toCompareValue < toBeComparedValue) return lessCaseColor;

  return defaultColor;
};

export const useAssetInputStyles = (
  currentValue: string,
  toCompare = 0,
): ((props?: React.CSSProperties) => ClassNameMap<'assetColor'>) => {
  return makeStyles((theme) => {
    const defaultColor = currentValue ? theme.palette.text.primary : theme.palette.text.secondary;
    const fontColor = getColorFromComparison({
      toCompareValue: Number(currentValue),
      toBeComparedValue: toCompare,
      greaterCaseColor: '#74D189',
      lessCaseColor: theme.palette.error.main,
      defaultColor,
    });

    return {
      assetColor: {
        color: fontColor,
      },
    };
  });
};
