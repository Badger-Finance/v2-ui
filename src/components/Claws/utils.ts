import BigNumber from 'bignumber.js';
import { INVALID_REASON } from './index';

export interface ValidationParams {
	amount: string | BigNumber;
	maximum: BigNumber;
	minimum?: BigNumber;
}

export function validateAmountBoundaries({
	amount,
	maximum,
	minimum = new BigNumber('0'),
}: ValidationParams): INVALID_REASON | undefined {
	let error: INVALID_REASON | undefined = undefined;

	const input = new BigNumber(amount);
	const amountExceedsBalance = input.gt(maximum);
	const amountIsLessThanMinimum = input.lt(minimum);

	if (amountIsLessThanMinimum) error = INVALID_REASON.UNDER_MINIMUM;
	if (amountExceedsBalance) error = INVALID_REASON.OVER_MAXIMUM;

	console.log({
		input: input.toString(),
		minimum: minimum.toString(),
		maximum: maximum.toString(),
		amountExceedsBalance,
		amountIsLessThanMinimum,
		error,
	});

	return error;
}
