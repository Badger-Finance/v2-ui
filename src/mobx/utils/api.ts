// api badger functions

// TODO: Update to api.badger.finance on v2 migration
const getApi = () => {
	if (process.env.NODE_ENV === 'production') {
		return 'https://api.sett.vision';
	}
	return 'https://hm77p4zo3l.execute-api.us-west-1.amazonaws.com/staging/';
};
const badgerApi = getApi();

export const getAssetsUnderManagement = (): any => {
	return fetch(`${badgerApi}/protocol/value?tokens=true`).then((response) => response.json());
};

// Geyser Data
export const getFarmData = async (): Promise<any> => {
	return await fetch(`${badgerApi}/protocol/farm`).then((response) => response.json());
};

export const getPpfs = async (): Promise<any> => {
	return await fetch(`${badgerApi}/protocol/ppfs`).then((response) => response.json());
};

export const getAssetPerformances = (setts: Array<any>): Promise<any> => {
	const performanceData = setts.map(async (sett: any) => {
		const assetKey = sett.asset.toLowerCase();
		return {
			...sett,
			asset: assetKey,
			title: sett.title,
			...(await getAssetPerformance(assetKey)),
		};
	});
	return Promise.all(performanceData);
};

const getAssetPerformance = async (asset: any): Promise<any> => {
	return await fetch(`${badgerApi}/protocol/sett/${asset}/performance`).then((response) => response.json());
};

export const getAssetPerformanceCharts = async (setts: Array<any>): Promise<any> => {
	const performanceData = setts.map(async (sett: any) => {
		const assetKey = sett.asset.toLowerCase();
		return {
			asset: assetKey,
			title: sett.title,
			data: await getPerformanceChartData(assetKey),
		};
	});
	return await Promise.all(performanceData);
};

const getPerformanceChartData = async (asset: any): Promise<any> => {
	return await fetch(`${badgerApi}/chart/sett/${asset}/performance?count=350`).then((response) => response.json());
};

export const getSettCharts = async (setts: Array<any>): Promise<any> => {
	const jarData = setts.map(async (sett) => {
		const assetKey = sett.asset.toLowerCase();
		return {
			title: sett.title,
			asset: assetKey,
			data: await getSettChartData(assetKey),
		};
	});
	return await Promise.all(jarData);
};

const getSettChartData = async (asset: any): Promise<any> => {
	return await fetch(`${badgerApi}/chart/sett/${asset}?count=1000`).then((response) => response.json());
};

export const getUserAccount = async (address: string): Promise<any> => {
	return await fetch(`${badgerApi}/protocol/earnings/${address}`).then((response) => response.json());
};

// api coingecko functions

const coingeckoApi = 'https://api.coingecko.com/api/v3';

export const getCoinData = async (coin: string): Promise<any> => {
	return await fetch(`${coingeckoApi}/coins/${coin}`).then((response) => response.json());
};

export const getEthPrice = async (coin: string): Promise<any> => {
	return await fetch(`${coingeckoApi}/coins/${coin}`)
		.then((response) => response.json())
		.then((jsonResponse) => jsonResponse['market_data']['current_price']['usd']);
};

// api util functions

export const formatUsd = (x: number): string => {
	try {
		const valueParts = x.toString().split('.');
		let value = valueParts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',');
		if (valueParts.length > 1) {
			value += '.' + valueParts[1].substring(0, 2);
		}
		return value;
	} catch (e) {
		console.log(e);
		return x.toString();
	}
};

export const formatWithCommas = (x: number): string => {
	try {
		const valueParts = x.toString().split('.');
		let value = valueParts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',');
		if (valueParts.length > 1) {
			value += '.' + valueParts[1].substring(0, 5);
		}
		return value;
	} catch (e) {
		console.log(e);
		return x.toString();
	}
};

// api badger functions

export const getBadgerRewards = async (data: string, address: string): Promise<any> => {
	return await fetch(
		`https://fzqm8i0owc.execute-api.us-east-1.amazonaws.com/prod/rewards/1/${data}/${address}`,
	).then((response) => response.json());
};
