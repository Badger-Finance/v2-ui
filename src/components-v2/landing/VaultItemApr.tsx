import { VaultDTO, VaultState } from '@badger-dao/sdk';
import { Box, Link, Typography } from '@material-ui/core';
import { makeStyles, Theme } from '@material-ui/core/styles';
import { isInfluenceVault } from 'components-v2/InfluenceVault/InfluenceVaultUtil';
import YieldBearingRewards from 'components-v2/YieldBearingVaults/YieldBearingRewards';
import { getYieldBearingVaultBySourceName } from 'components-v2/YieldBearingVaults/YieldBearingVaultUtil';
import { FLAGS } from 'config/environment';
import { useVaultInformation } from 'hooks/useVaultInformation';
import { StoreContext } from 'mobx/stores/store-context';
import { numberWithCommas } from 'mobx/utils/helpers';
import React, { MouseEvent, useContext, useState } from 'react';

import VaultApyInformation from '../VaultApyInformation';

const useStyles = makeStyles((theme: Theme) => ({
  root: {
    cursor: 'pointer',
  },
  apr: {
    cursor: 'default',
    fontSize: 16,
  },
  apyInfo: {
    marginLeft: 5,
  },
  aprDisplay: {
    justifyContent: 'flex-end',
    [theme.breakpoints.down('sm')]: {
      justifyContent: 'flex-start',
    },
  },
  projectedApr: {
    fontSize: 10,
    marginTop: 3,
    color: '#FFFFFF99',
  },
  yieldBearingRewards: {
    fontSize: 12,
    marginTop: 5,
    width: '100%',
    '& a': {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'flex-end',
      [theme.breakpoints.down('sm')]: {
        justifyContent: 'flex-start',
      },
      '& img': {
        marginRight: 5,
      },
    },
  },
}));

interface Props {
  vault: VaultDTO;
  isDisabled?: boolean;
}

const VaultItemApr = ({ vault }: Props): JSX.Element => {
  const classes = useStyles();
  const [showApyInfo, setShowApyInfo] = useState(false);
  const [openYieldBearingRewardsModal, setOpenYieldBearingRewardsModal] = useState(false);
  const { projectedVaultBoost, vaultBoost } = useVaultInformation(vault);

  const store = useContext(StoreContext);
  const { vaults } = store;
  const yieldSourcesAprTotal = vault.sourcesApy.reduce((max, source) => {
    const yieldVault = getYieldBearingVaultBySourceName(source.name);
    if (yieldVault !== undefined) {
      const current = vaults.getVault(yieldVault.vaultId)?.apy ?? 0;
      if (current > max) {
        max = current;
      }
    }
    return max;
  }, 0);

  const handleApyInfoClick = (event: MouseEvent<HTMLElement>) => {
    event.stopPropagation();
    setShowApyInfo(true);
  };

  const handleYieldBearingRewardsClick = (event: MouseEvent<HTMLElement>) => {
    event.stopPropagation();
    setOpenYieldBearingRewardsModal(true);
    return false;
  };

  const handleClose = () => {
    setShowApyInfo(false);
  };

  if (!vault.apr) {
    return (
      <Typography className={classes.apr} variant="body1" color={'textPrimary'}>
        --%
      </Typography>
    );
  }

  const isNewVault = vault.state === VaultState.Experimental || vault.state === VaultState.Guarded;
  const aprDisplay = isNewVault ? (
    <>
      <img src={'assets/icons/new-vault.svg'} alt="New Vault" /> New Vault
    </>
  ) : (
    `${numberWithCommas(vaultBoost.toFixed(2))}%`
  );
  const isInfluence = isInfluenceVault(vault.vaultToken);

  return (
    <Box
      display="flex"
      alignItems={FLAGS.APY_EVOLUTION ? 'flex-end' : 'flex-start'}
      flexDirection="column"
      className={classes.root}
      onClick={(e) => e.stopPropagation()}
    >
      <Box className={classes.aprDisplay} display="flex" width="100%" onClick={handleApyInfoClick}>
        <Typography variant={isNewVault ? 'subtitle1' : 'body1'} color={'textPrimary'} display="inline">
          {aprDisplay}
        </Typography>
        <img src="/assets/icons/apy-info.svg" className={classes.apyInfo} alt="apy info icon" />
      </Box>
      {FLAGS.APY_EVOLUTION && yieldSourcesAprTotal > 0 && (
        <Box className={classes.yieldBearingRewards}>
          <Link color="primary" onClick={handleYieldBearingRewardsClick}>
            <img width="9" src="assets/icons/yield-bearing-rewards.svg" alt="Yield-Bearing Rewards" /> Yield-Bearing
            Rewards.
          </Link>
          <Typography onClick={handleApyInfoClick} variant="inherit">
            Rewards earn up to {yieldSourcesAprTotal.toFixed(2)}%
          </Typography>
        </Box>
      )}
      {!FLAGS.APY_EVOLUTION && !isInfluence && projectedVaultBoost !== null && (
        <Box display="flex" onClick={handleApyInfoClick}>
          <Typography className={classes.projectedApr}>
            Current: {`${numberWithCommas(projectedVaultBoost.toFixed(2))}%`}
          </Typography>
        </Box>
      )}
      <VaultApyInformation
        open={showApyInfo}
        vault={vault}
        boost={vaultBoost}
        projectedBoost={projectedVaultBoost}
        onClose={handleClose}
      />
      <YieldBearingRewards
        open={openYieldBearingRewardsModal}
        onModalClose={() => setOpenYieldBearingRewardsModal(false)}
      />
    </Box>
  );
};

export default VaultItemApr;
