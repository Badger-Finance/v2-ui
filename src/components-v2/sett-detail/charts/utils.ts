import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { Sett } from '../../../mobx/model/setts/sett';
import { fetchSettChartInformation } from '../../../mobx/utils/apiV2';
import { SettSnapshotGranularity } from '../../../mobx/model/setts/sett-snapshot';
import { SettChartData } from '../../../mobx/model/setts/sett-charts';

dayjs.extend(utc);

export enum SettChartTimeframe {
	'day' = 1,
	'week' = 7,
	'month' = 30,
}

/**
 * Fetches chart information since the provided timeframe until the current date
 * @param sett
 * @param timeframe
 */
export const fetchSettChart = async (sett: Sett, timeframe: SettChartTimeframe): Promise<SettChartData[] | null> => {
	const timeframeDays = 1 * timeframe;
	const isDayTimeFrame = timeframe === SettChartTimeframe.day;

	// if timeframe is just one day then we want the granularity to be hours
	const granularity = isDayTimeFrame ? SettSnapshotGranularity.HOUR : SettSnapshotGranularity.DAY;

	const now = dayjs().utc(); // query until current date
	const from = dayjs(now).subtract(timeframeDays, 'days').utc();

	const fetchedData = await fetchSettChartInformation(sett.vaultToken, from.toDate(), now.toDate(), granularity);

	if (!fetchedData) {
		return null;
	}

	// data needs to be ascending sorted
	return fetchedData
		.sort((a, b) => a.timestamp - b.timestamp)
		.map((d) => ({ ...d, timestamp: new Date(d.timestamp) }));
};
