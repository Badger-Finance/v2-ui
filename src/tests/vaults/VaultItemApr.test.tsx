import { BouncerType, Protocol, VaultBehavior, VaultDTO, VaultState, VaultType, VaultVersion } from '@badger-dao/sdk';
import React from 'react';
import { SAMPLE_VAULT } from 'tests/utils/samples';

import VaultItemApr from '../../components-v2/landing/VaultItemApr';
import { checkSnapshot } from '../utils/snapshots';

const normalVault: VaultDTO = {
  available: 0,
  name: 'Badger',
  asset: 'Badger',
  vaultAsset: 'bBadger',
  state: VaultState.Open,
  underlyingToken: '0x0000000000000000000000000000000000000001',
  vaultToken: '0x0000000000000000000000000000000000000001',
  value: 10000,
  balance: 100,
  protocol: Protocol.Badger,
  pricePerFullShare: 0.18907615705168573,
  tokens: [
    {
      address: '0x0000000000000000000000000000000000000001',
      name: 'Badger',
      symbol: 'BADGER',
      decimals: 10,
      balance: 100,
      value: 10000,
    },
  ],
  apr: 8.174287821972374,
  apy: 8.174287821972374,
  boost: {
    enabled: false,
    weight: 0,
  },
  sources: [
    {
      name: 'Vault Compounding',
      apr: 8.174287821972374,
      boostable: false,
      minApr: 8.174287821972374,
      maxApr: 8.174287821972374,
    },
  ],
  sourcesApy: [
    {
      name: 'Vault Compounding',
      apr: 8.174287821972374,
      boostable: false,
      minApr: 8.174287821972374,
      maxApr: 8.174287821972374,
    },
  ],
  bouncer: BouncerType.None,
  strategy: {
    address: '0x4a8651F2edD68850B944AD93f2c67af817F39F62',
    withdrawFee: 0,
    performanceFee: 0,
    strategistFee: 0,
    aumFee: 0,
  },
  type: VaultType.Native,
  version: VaultVersion.v1,
  behavior: VaultBehavior.Compounder,
  yieldProjection: {
    yieldApr: 0,
    yieldPeriodApr: 0,
    yieldTokens: [],
    yieldTokensPerPeriod: [],
    yieldValue: 0,
    harvestApr: 0,
    harvestApy: 0,
    harvestPeriodApr: 0,
    harvestPeriodApy: 0,
    harvestTokens: [],
    harvestTokensPerPeriod: [],
    harvestValue: 0,
  },
  lastHarvest: Date.now(),
};

describe('VaultItemApr', () => {
  describe('No APR Vaults', () => {
    it('renders zero APR', () => {
      checkSnapshot(<VaultItemApr boost={0} vault={{ ...normalVault, apr: 0 }} />);
    });
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  describe('Boosted Vaults', () => {
    const mockUserBoost = 10;
    it('displays correct APR and boost information', () => {
      checkSnapshot(<VaultItemApr vault={SAMPLE_VAULT} boost={mockUserBoost} />);
    });
  });
});
