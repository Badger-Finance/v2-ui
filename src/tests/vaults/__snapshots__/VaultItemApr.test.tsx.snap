// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VaultItemApr Boosted Vaults displays correct APR and boost information 1`] = `
<div>
  <div
    class="MuiBox-root-5 MuiBox-root-6 makeStyles-root-1"
  >
    <div
      class="MuiBox-root-5 MuiBox-root-7"
    >
      <p
        class="MuiTypography-root-8 MuiTypography-body1-10 MuiTypography-colorTextPrimary-33 MuiTypography-displayInline-36"
      >
        10.00%
      </p>
      <img
        alt="apy info icon"
        class="makeStyles-apyInfo-3"
        src="/assets/icons/apy-info.svg"
      />
    </div>
    <div
      class="MuiBox-root-5 MuiBox-root-38"
    >
      <p
        class="MuiTypography-root-8 makeStyles-projectedApr-4 MuiTypography-body1-10"
      >
        Proj. 
        20.00%
      </p>
    </div>
  </div>
</div>
`;

exports[`VaultItemApr No APR Vaults renders zero APR 1`] = `
<div>
  <p
    class="MuiTypography-root-5 makeStyles-apr-2 MuiTypography-body1-7 MuiTypography-colorTextPrimary-30"
  >
    --%
  </p>
</div>
`;
